### DISCUSSION TITLE
Topic #2: Case Study: Drasil for Understanding Chunk/KF Changes and Assisted Updates for Instances

### DISCUSSION THREAD

**Discussion by balacij** [JacquesCarette/Drasil#3007] [Ideas]
# Problem

For unrelated reasons, refactoring code to update against newer versions of
software dependencies is a regular task developers take on. Typically, an
upgrade path is understood to developers of the software dependencies, but that
information is lost on them. Thus, unfortunately, updating code often requires
reading a manually written changelog (which usually don't explain enough of the
nitty-gritty details), comparing API documentation, and a lot of guess-and-check
work to ensure software remains stable, and no unexpected changes occur (e.g.,
bugs, business logic changes, etc.).

# Purpose

Through formalizing how chunk types and code were altered, the purpose of this
research is to investigate how building libraries and executable software with
Drasil can allow for assisted dependency updates (though doubtful, potentially
also "fully automated").

# Research Questions

Can we...
1. build and share libraries for Drasil programs?
2. formalize how source code changes for various languages?
3. formalize how source code changes affect dependent software?

This would likely rely largely on having Drasil understand the types it frequently works with.

**Comment by smiths** [comment]
I like this trend of capturing ideas in the Discussion.  :smile:

I agree with that the problem of newer versions of software dependencies, but I don't understand the proposed role for Drasil.  In this scenario are the libraries that change libraries generated by Drasil, or are they libraries that are external to Drasil?

**Reply by balacij** [reply]
I'm glad! I think GitHub Discussions is good for these, especially. However, I'm not quite sure how much we want to turn it into a general "online forum".

I never really thought about things external, but I don't see why those wouldn't be included in this. I was primarily focused on the "drasil knowledge libraries", such as `drasil-lang`, `drasil-theory`, etc, and the types and data included in them. So, I think that by extension, the generated library artifacts would also be updated, and that changes to internal representations of external software libraries could affect what we generate. The general idea is focused on understanding: (a) how types of knowledge fragments are changed, (b) how said changes would affect the instances of those knowledge fragments, and (c) how we can formalize those changes such that we can automatically (where possible) or provide assistance in updating the data.

This project _might_ rely on #3003 because we would need to work with Drasils related types, but we might not want to delve into reflection.

**Reply by smiths** [reply]
Thank you @balacij.  That makes sense.

**Comment by balacij** [comment]
The title might need to be re-worded, perhaps: "Topic https://github.com/JacquesCarette/Drasil/issues/2: Case Study: Drasil for Understanding Chunk/KF Changes and Assisted Updates for Instances"

**Reply by smiths** [reply]
Yes, I like that proposed title.  The old title made me think that this was around issues like how we deal with the Boost library, and other external dependencies.

**Comment by JacquesCarette** [comment]
I think it would be worthwhile to rewrite this given the discussion above and the discussion in our meeting.

These topics most likely should be written in a way to be understandable to people outside Drasil (and the least the gist of them). And should contain concrete examples to really illustrate the problem.
